src/DotNetty.Common/ThreadLocalPool.cs

```diff
@@ -41,22 +51,296 @@ public bool Release<T>(T value)
             }
         }
 
-        internal sealed class Stack : Stack<Handle>
+        internal sealed class WeakOrderQueue
+        {
+            const int LinkCapacity = 16;
+
+            sealed class Link
+            {
+                internal int readIndex;
+                internal int writeIndex;
+
+                internal readonly Handle[] elements;
+                internal Link next;
+
+                internal Link()
+                {
+                    this.elements = new Handle[LinkCapacity];
+                }
+            }
+
+            Link head, tail;
+            internal WeakOrderQueue next;
+            internal WeakReference<Thread> ownerThread;
+
+            internal bool IsEmpty
+            {
+                get { return this.tail.readIndex == Volatile.Read(ref this.tail.writeIndex); }
+            }
+
+            internal WeakOrderQueue(Stack stack, Thread thread)
+            {
+                Contract.Requires(stack != null);
+
+                this.ownerThread = new WeakReference<Thread>(thread);
+                this.head = this.tail = new Link();
+                lock (stack)
+                {
+                    this.next = Volatile.Read(ref stack.headQueue);
+                    Volatile.Write(ref stack.headQueue, this);
+                }
+            }
+
+            internal void Add(Handle handle)
+            {
+                Contract.Requires(handle != null);
+
+                //handle.lastRecycledId = id;
+
+                Link tail = this.tail;
+                int writeIndex = Volatile.Read(ref tail.writeIndex);
+                if (writeIndex == LinkCapacity)
+                {
+                    this.tail = tail = tail.next = new Link();
+                    writeIndex = Volatile.Read(ref tail.writeIndex);
+                }
+                tail.elements[writeIndex] = handle;
+                //handle.Stack = null;
```

Comment:
why is it commented out? If it's because `Handle.Stack` is `readonly` - remove the `readonly`.